boot/bootutil/src/loader.c

```diff
@@ -1583,6 +1586,16 @@ context_boot_go(struct boot_loader_state *state, struct boot_rsp *rsp)
         /* Set the previously determined swap type */
         bs.swap_type = BOOT_SWAP_TYPE(state);
 
+#if defined(MCUBOOT_OVERWRITE_ONLY) && defined(MCUBOOT_DOWNGRADE_PREVENTION)
+        /* Check if version of secondary is sufficient */
+        rc = boot_is_version_sufficient(
+                &boot_img_hdr(state, BOOT_PRIMARY_SLOT)->ih_ver,
+                &boot_img_hdr(state, BOOT_SECONDARY_SLOT)->ih_ver);
+        if (rc != 0) {
+                BOOT_LOG_WRN("insufficient version");
+                goto out;
+        }
+#endif
```

Comment:
I'm not sure if this version check is happening in the right stage (after image validation and swap type determination), and also it may requires further actions depending on the swap type.
In it's current form if we have valid images in both the primary and secondary slots, but the image in the secondary slot has an older version number the `boot_is_version_sufficient` will return with an error an the code will jump to the `out` flag. In this case the device won't boot any image.
I would suggest the move this check into the `boot_validated_swap_type` or `boot_validate_slot` functions (it should be checked against the secondary slot's image only). In this case if the version number comparison fails it would simply set the swap type to `NONE` or `FAIL`, erase the contect if the secondary slot and then boot the previously used image from the  primary slot (without performing an upgrade). Basically I would just make this check part of the image validation process.

Reply:
I moved the check now. It looks a bit bloated, but I want to perform the version check before doing the signature check, and it was not easy to cram it in to the existing if checks without making them look weird. Feedback and proposals is, as always, welcome.